core/kernel/generic_pm.c

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018, Linaro Limited
+ */
+
+#include <keep.h>
+#include <kernel/generic_pm.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+#include <types_ext.h>
+
+/*
+ * Generic power management support for platforms and drivers.
+ * Core components can register callback for suspend and resume
+ * sequences. Callback can be related to a specific subsystem,
+ * based on an opaque identifier. Callback has some ordering
+ * directives as specified by the PM_CB_LEVEL_xxx value.
+ *
+ * Platform specific power sequence can call pm_suspend_resume()
+ * for a suspend or resume operation and for a target subsystem.
+ * Registered callbacks for that subsystem are called with the
+ * ordering directive.
+ */
+
+static bool __maybe_unused cb_is_valid(int (*cb)(enum pm_op op, void *hdl),
+				       void *hdl)
+{
+	paddr_t cb_phy = virt_to_phys(cb);
+	paddr_t hdl_phy = virt_to_phys(hdl);
+	bool __maybe_unused valid;
+
+	valid = (phys_to_virt(cb_phy, MEM_AREA_TEE_RAM_RX) == cb) &&
+		(!hdl_phy ||
+		 (phys_to_virt(hdl_phy, MEM_AREA_TEE_RAM_RX) == hdl) ||
+		 (phys_to_virt(hdl_phy, MEM_AREA_TEE_RAM_RO) == hdl) ||
+		 (phys_to_virt(hdl_phy, MEM_AREA_TEE_RAM_RW) == hdl));
+
+	if (!valid) {
+		EMSG("pm_cb mandates unpaged arguments %p %p", (void *)cb, hdl);
+	}
+
+	return valid;
+}
+
+static bool __maybe_unused cb_level_is_valid(enum pm_op op)
+{
+	switch (op) {
+	case PM_OP_SUSPEND:
+	case PM_OP_RESUME:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static bool __maybe_unused pm_subsys_is_valid(unsigned int pm_subsys)
+{
+	return pm_subsys <= UINT16_MAX;
+}
+
+struct pm_cb {
+	int (*callback)(enum pm_op op, void *handle);
```

Comment:
The callback should probably get all the other elements of this struct as arguments.

Reply:
yes probably. I think the callback proto will even change since comments from @cneveux.